dagjose/dagjose_assembler.go

```diff
@@ -2,161 +2,166 @@ package dagjose
 
 import (
 	"fmt"
-
 	"github.com/ipfs/go-cid"
 	ipld "github.com/ipld/go-ipld-prime"
 	basicnode "github.com/ipld/go-ipld-prime/node/basic"
 	"github.com/ipld/go-ipld-prime/node/mixins"
+	"github.com/pkg/errors"
 )
 
-var (
-	_ ipld.Node          = dagJOSENode{}
-	_ ipld.NodePrototype = &DagJOSENodePrototype{}
-	_ ipld.NodeAssembler = &dagJOSENodeBuilder{}
+type (
+	dagJOSENodePrototype struct{}
+	maState              uint
+)
+
+const (
+	maStateInitial     maState = iota // also the 'expect key or finish' state
+	maStateMidKey                     // waiting for a 'finished' state in the KeyAssembler.
+	maStateExpectValue                // 'AssembleValue' is the only valid next step
+	maStateMidValue                   // waiting for a 'finished' state in the ValueAssembler.
+	maStateFinished                   // finished
 )
 
-type DagJOSENodePrototype struct{}
+var (
+	dagJOSEMixin = mixins.MapAssembler{TypeName: "dagJOSE"}
+)
 
-func (d *DagJOSENodePrototype) NewBuilder() ipld.NodeBuilder {
-	return &dagJOSENodeBuilder{dagJose: DagJOSE{}}
+func (d *dagJOSENodePrototype) NewBuilder() ipld.NodeBuilder {
+	return &dagJOSENodeBuilder{dagJOSE: DAGJOSE{}}
 }
 
-// Returns an instance of the DagJOSENodeBuilder which can be passed to
-// ipld.Link.Load and will build a dagjose.DagJOSE object. This should only be
-// neccesary in reasonably advanced situations, most of the time you should be
+// NewBuilder Returns an instance of the DagJOSENodeBuilder which can be passed to
+// ipld.Link.Load and will build a dagjose.DAGJOSE object. This should only be
+// necessary in reasonably advanced situations, most of the time you should be
 // able to use dagjose.LoadJOSE
 func NewBuilder() ipld.NodeBuilder {
-	return &dagJOSENodeBuilder{dagJose: DagJOSE{}}
+	return &dagJOSENodeBuilder{dagJOSE: DAGJOSE{}}
 }
 
-type maState uint8
-
-const (
-	maState_initial     maState = iota // also the 'expect key or finish' state
-	maState_midKey                     // waiting for a 'finished' state in the KeyAssembler.
-	maState_expectValue                // 'AssembleValue' is the only valid next step
-	maState_midValue                   // waiting for a 'finished' state in the ValueAssembler.
-	maState_finished                   // finised
-)
-
-// An implementation of `ipld.NodeBuilder` which builds a `dagjose.DagJOSE`
+// An implementation of `ipld.NodeBuilder` which builds a `dagjose.DAGJOSE`
 // object. This builder will throw an error if the IPLD data it is building
 // does not match the schema specified in the spec
 type dagJOSENodeBuilder struct {
-	dagJose DagJOSE
+	dagJOSE DAGJOSE
 	state   maState
 	key     *string
 }
 
-var dagJoseMixin = mixins.MapAssembler{TypeName: "dagjose"}
-
-func (d *dagJOSENodeBuilder) BeginMap(sizeHint int64) (ipld.MapAssembler, error) {
-	if d.state != maState_initial {
-		panic("misuse")
```

Comment:
This `panic` is used correctly. If this gets run, that means something majorly bad happened outside the scope of this DAG-JOSE implementation. And in order to ensure bad data doesn't get read or written to an IPFS node, it should `panic`.

Reply:
Well you can recover from a panic (https://golang.org/doc/effective_go#recover and https://go.dev/blog/defer-panic-and-recover), but when APIs are misused, you don't want to encourage recovery, especially when it could cause loss of user data. Also, when it comes to IPFS plugins that implement a IPLD codec, each time the node assemblers are run are in independent goroutines. A panic crashes the goroutine that calls it not the entire Go process.

Reply:
I do agree with you, it seems like this is a good case for panic since the library intentionally does not allow for an error to be returned. I will update -- thanks for the links.